### PR TITLE
Allow cucumber to be used as an unreleased (git) dependency

### PR DESCRIPTION
**Pull request by aslakhellesoy**
We sometimes want to use cucumber with features that haven't been released.

Adding `"cucumber": "cucumber/cucumber#some-branch"` to `package.json` will cause npm/yarn to pull from git, but it will *not* run babel over the files.

To make this work I've created a script that builds cucumber (if it's not already built), and trigger this in `postinstall`. I've also changed `prepublish` to use this script, to prevent babel from running twice.

This technique is already used in cucumber-expressions, and has been working nicely.

### CODE REVIEW

**Comment by kriswill** [comment]
This can be super helpful for people wanting to validate pre-releases, branches and PRs with their existing projects.  I just tried installing off of github:cucumber/cucumber-js#v3.0.1, with no luck.  It's missing `/lib`

**Comment by aslakhellesoy** [comment]
Yeah I know it's buggy. It's supposed to run babel to build files in lib, but I haven't managed to make it work when it's installed as a dependency.

I wonder if it would be ok to add the babelised code to git. A bit ugly, but a more pragmatic approach maybe? WDYT @charlierudolph or @jbpros ?

File: package.json [hunk 1]
```diff
@@ -194,8 +194,9 @@
     "lint-code": "eslint \"{example,features,scripts,src,test}/**/*.js\"",
     "lint-autofix": "eslint --fix \"{example,features,scripts,src,test}/**/*.js\"",
     "lint": "yarn run lint-code && yarn run lint-dependencies",
-    "prefeature-test": "yarn run build-local",
-    "prepublish": "yarn run build-local",
+    "postinstall": "node scripts/build-local.js",
+    "prefeature-test": "node scripts/build-local.js",
+    "prepublish": "node scripts/build-local.js",
```
Comment (charlierudolph) [inline package.json]: prepublish and prefeature-test should remain as they were. The lib folder needs to be updated even it if exists in order to ensure it contains the latest code

File: package.json [hunk 2]
```diff
@@ -194,8 +194,9 @@
     "lint-code": "eslint \"{example,features,scripts,src,test}/**/*.js\"",
     "lint-autofix": "eslint --fix \"{example,features,scripts,src,test}/**/*.js\"",
     "lint": "yarn run lint-code && yarn run lint-dependencies",
-    "prefeature-test": "yarn run build-local",
-    "prepublish": "yarn run build-local",
+    "postinstall": "node scripts/build-local.js",
```
Comment (charlierudolph) [inline package.json]: Thoughts on calling the script something like `node scripts/build-non-release.js`?

File: scripts/build-local.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+var path = require('path')
+var fs = require('fs')
+var spawn = require('child_process').spawn
+
+// Runs `npm run build-local` if the lib dir is absent
```
Comment (charlierudolph) [inline scripts/build-local.js]: This script does more than what the comment says. I think it should be something like, "If installed as an unreleased git dependency, installs the dev dependencies and compiles the source"

File: scripts/build-local.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+fs.access(path.join(__dirname, '..', 'lib'), function(err) {
+  if (err) {
+    // If we're installed as a dependency in another project, install our own node_modules,
+    // which are required to run build-local
+    fs.access(path.join(__dirname, '..', 'node_modules', 'babel-plugin-external-helpers'), function(err) {
```
Comment (charlierudolph) [inline scripts/build-local.js]: I think this check can be removed and we instead always run `npm install`

**Comment by charlierudolph** [comment]
I'm fine with this script and do not want to add the babelised code to git. @aslakhellesoy I don't think @kriswill was using the code introduced in this PR and was just saying it would be nice to have. We should validate that we can install this branch as a git dependency before merging

File: scripts/build-local.js (hunk 4 above)
Comment (aslakhellesoy) [reply scripts/build-local.js]: No, it needs to be there, or we'll end up with infinite recursion.

After the script runs `npm install`, the `postinstall` script runs again, and this time we *don't* want to run `npm install`.

**Comment by aslakhellesoy** [comment]
Thanks for all the feedback @charlierudolph - I think I've dealt with it all.

File: scripts/build-non-release.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+var path = require('path')
+var fs = require('fs')
+var spawn = require('child_process').spawn
+
+// If installed as an unreleased git dependency, installs the dev dependencies
+// and compiles the source. Uses yarn by default. Set `CUCUMBER_UNRELEASED_BUILD_TOOL=npm`
+// if you prefer to use npm (it's slower).
+
+const buildTool = process.env.CUCUMBER_UNRELEASED_BUILD_TOOL || 'yarn'
```
Comment (kriswill) [inline scripts/build-non-release.js]: NPM will always be present with node.js, yarn isn't guaranteed to be present.  Perhaps, test if yarn is installed before using it?

Comment (kriswill) [reply scripts/build-non-release.js]: Here's the timings with latest NPM:
```
kris git[master?] ~/src/oss/cucumber-js  
→ npm --versions
{ cucumber: '3.0.1',
  npm: '5.3.0',
  ares: '1.10.1-DEV',
  cldr: '31.0.1',
  http_parser: '2.7.0',
  icu: '59.1',
  modules: '57',
  nghttp2: '1.22.0',
  node: '8.4.0',
  openssl: '1.0.2l',
  tz: '2017b',
  unicode: '9.0',
  uv: '1.13.1',
  v8: '6.0.286.52',
  zlib: '1.2.11' }
```
and installation:
```
kris git[master] ~/src/oss/cucumber-js  
→ npm I
...
npm notice created a lockfile as package-lock.json. You should commit this file.
added 912 packages in 24.784s
```

Comment (kriswill) [reply scripts/build-non-release.js]: without yarn installed globally:
```
> cucumber-expressions@4.0.3 postinstall /Users/kris/src/oss/cucumber-js/node_modules/cucumber-expressions
> node scripts/postinstall.js

npm WARN prepublish-on-install As of npm@5, `prepublish` scripts are deprecated.
npm WARN prepublish-on-install Use `prepare` for build steps and `prepublishOnly` for upload-only.
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> cucumber@3.0.1 prepublish /Users/kris/src/oss/cucumber-js
> yarn run build-local

sh: yarn: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! cucumber@3.0.1 prepublish: `yarn run build-local`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the cucumber@3.0.1 prepublish script.
```

Comment (kriswill) [reply scripts/build-non-release.js]: Also, using shell scripts during an npm install will make the whole project not work on Windows.  Not sure if that is important to you guys.

Comment (kriswill) [reply scripts/build-non-release.js]: https://github.com/charlike/detect-installed seems like a reasonable module to use to detect the presence of yarn:

```js
const detectInstalled = require('detect-installed')
const exists = await detectInstalled('yarn')
```

**Comment by charlierudolph** [comment]
When testing this it is working when installing  via `yarn` but hitting an error when installing via npm. I believe this caused by the fact that npm still uses the `files` field in the package.json on git installs and strips out the `src` and `script` folders. I'd prefer not to open up the `files` in order to avoid sending the `src` and `scripts` with each install. However, I currently don't know what else we could do.

**Comment by aslakhellesoy** [comment]
I suggest we make yarn a requirement then, and print an error message saying it's required if they don't have it.

People on the edge ude yarn anyway.

I don't think this will affect more than a handful of people.

**Comment by kriswill** [comment]
IMHO, The only problem with requiring yarn for installing the module, is yarn isn't a natural dependency, unless you include it in your package.json, which I don't think is a good pattern to include a package management system in a single module.  Most people using node, do not use yarn, and this only serves to complicate the procedure for getting up and running.

**Comment by charlierudolph** [comment]
Since this doesn't have a reliable way for us to do this (without doubling the size of the published package), I'd prefer we just close this and focus on releasing often. The release process is pretty simple and anyone with the commit bit can do it.

**Comment by aslakhellesoy** [comment]
Fair enough, we'll just keep these scripts on a branch and use them when we need it. It works for us.

**Pull request closed without merging**